keps/sig-cluster-lifecycle/0032-create-a-k8s-io-component-repo.md

```diff
@@ -0,0 +1,245 @@
+
+### Part 3: HTTPS serving
+
+Many Kubernetes controllers are clients to the API server and run as daemons. In order to expose information on how the component is doing (e.g. profiling, metrics, current configuration, etc.), an HTTPS server is run.
+
+#### Common endpoints
+
+In order to make it easy to expose this kind of information, a package is made in this new repo that hosts this common code. Initially targeted endpoints are "/version," "/healthz," "/configz," "/pprof," and "/metrics."
+
+#### Standardized authentication / authorization
+
+In order to not expose this kind of information (e.g. metrics) to anyone that can talk to the component, it may utilize SubjectAccessReview requests to the API server, and hence delegate authentication and authorization to the API server. It should be easy to add this functionality to your component.
+
+### Part 4: Sample implementation in k8s.io/sample-component
+
+Provides an example usage of the three main functions of the _k8s.io/component_ repo, implementing ComponentConfig, the CLI wrapper tooling and the common HTTPS endpoints with delegated auth.
+
+### Code structure
+
+- k8s.io/component
+  - config/
+    - Would hold internal, shared ComponentConfig types across core components
+    - {v1,v1beta1,v1alpha1}
+      - Would hold external, shared ComponentConfig types across core components
+    - serializer/
+      - Would hold common methods for encoding/decoding ComponentConfig
+    - testing/
+      - Would hold common testing code for use in unit tests local to the implementation of ComponentConfig.
+  - cli/
+    - Would hold common methods and types for building a k8s component command (building on top of github.com/spf13/{pflag,cobra})
+    - options/
+      - Would hold flag definitions
+    - testing/
+      - Would hold common testing code for use in unit tests local to the implementation of the code
+    - logging/
+      - Would hold common code for using _k8s.io/klog_
+  - server/
+    - auth/
+      - Would hold code for implementing delegated authentication and authorization to Kubernetes
+    - configz/
+      - Would hold code for implementing a `/configz` endpoint in the component
+    - healthz/
+      - Would hold code for implementing a `/healthz` endpoint in the component
+    - metrics/
+      - Would hold code for implementing a `/metrics` endpoint in the component
+    - pprof/
+      - Would hold code for implementing a `/pprof` endpoint in the component
+    - version/
+      - Would hold code for implementing a `/version` endpoint in the component
+
+### Timeframe and Implementation Order
+
+**Objective:** The ComponentConfig part done for v1.14
+
+**Stretch goal:** Get the CLI and HTTPS server parts done for v1.14.
+
+**Implementation order:**
+
+1. Create the k8s.io/component repo with the initial ComponentConfig shared code
+2. Move shared v1alpha1 ComponentConfig types and references from `k8s.io/api{server,machinery}/pkg/apis/config` to `k8s.io/component/config`
```

Comment:
And agree on what should be shared :)

Reply:
Yeah, right. So far we have three types in total in these two k8s.io/apimachinery & k8s.io/apiserver packages that the core components are using, so first just move these.